tests/utils_tests/test_functional.py

```diff
@@ -265,3 +265,93 @@ def foo(cls):
+        with self.assertRaisesMessage(TypeError, "takes 1 positional argument but 2 were given"):
+            keep_lazy_single_argument("test", "testing")
+
+    def test_keep_lazy_single_argument_given_multiple_mixed_arguments(self):
+        with self.assertRaisesMessage(TypeError, "got an unexpected keyword argument 'val'"):
+            keep_lazy_single_argument("test", val="testing")
+
+    def test_keep_lazy_single_argument_given_multiple_keyword_arguments(self):
+        with self.assertRaisesMessage(TypeError, "got an unexpected keyword argument 'val'"):
+            keep_lazy_single_argument(value="test", val="testing")
+
+    def test_keep_lazy_multiple_positional_arguments(self):
+        self.assertEqual(
+            keep_lazy_multiple_arguments("test1", "test2", "test3"),
+            ("test1", "test2", "test3")
+        )
+
+    def test_keep_lazy_multiple_keyword_arguments(self):
+        self.assertEqual(
+            keep_lazy_multiple_arguments(val1="test1", val2="test2", val3="test3"),
+            ("test1", "test2", "test3")
+        )
+
+    def test_keep_lazy_multiple_invalid_keyword_argument(self):
+        with self.assertRaisesMessage(TypeError, "got an unexpected keyword argument 'val'"):
+            keep_lazy_multiple_arguments(val="test")
+
+    def test_keep_lazy_multiple_given_too_few_positional_arguments(self):
+        with self.assertRaisesMessage(TypeError, "missing 1 required positional argument: 'val3'"):
+            keep_lazy_multiple_arguments("test", "testing")
+
+    def test_keep_lazy_multiple_given_too_few_keyword_arguments(self):
+        with self.assertRaisesMessage(TypeError, "missing 1 required positional argument: 'val3'"):
+            keep_lazy_multiple_arguments(val1="test", val2="testing")
+
+    def test_keep_lazy_multiple_given_too_many_positional_arguments(self):
+        with self.assertRaisesMessage(TypeError, "takes 3 positional arguments but 4 were given"):
+            keep_lazy_multiple_arguments("test1", "test2", "test3", "test4")
+
+    def test_keep_lazy_multiple_given_too_many_keyword_arguments(self):
+        with self.assertRaisesMessage(TypeError, "got an unexpected keyword argument 'val4'"):
+            keep_lazy_multiple_arguments(val1="test1", val2="test2", val3="test3", val4="test4")
+
+    def test_posargs(self):
+        """
+        A function which accepts *args must be dispatched to the multiple argument wrapper
+        """
+        self.assertEqual(
+            keep_lazy_posargs("this", "may", "take", "many", "arguments"),
+            ("this", "may", "take", "many", "arguments")
+        )
+
+    def test_kwargs(self):
+        """
+        A function which accepts **kwargs must be dispatched to the multiple argument wrapper
+        """
+        self.assertEqual(
+            keep_lazy_kwargs(this="may", take="many", arguments="too"),
+            {"this": "may", "take": "many", "arguments": "too"}
+        )
```

Comment:
So, these two correctly dispatch to the `multiple_argument` form due to the check `if v.kind not in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD)` in the decorator.

Buuuut, removing that line doesn't cause the tests to fail. They just go through the single argument version and incorrectly pass because it passes down `*args/**kwargs` regardless, and _only_ checked the first argument for being a `Promise` in the process.

So... not ideal, probably needs a better test than I can figure out right now.

Reply:
Yeah, so none of the tests failed if I forced either the multiple or single argument form of the function. That can't be right.

Reply:
Yeah, I noted as much on the ticket, in a roundabout way. TBF they're kind of placeholders, because I dunno off the top of my head how to demonstrate which (internal only) wrapper they went through (being that the whole point is the wrappers should be transparent), and that the right thing happened. I mostly added the tests and then a bunch of print calls in the wrappers to check they went through the right ones, and didn't raise or swallow exceptions.

I'd welcome any ideas for testing everything does the right thing - there's not much point in my continuing to iterate (eg: with some of the changes you proposed) until the tests can properly validate something more than they currently do.